[docs] Add `Form` integration page (#2989)
mui/base-ui: docs/src/app/(public)/(content)/react/handbook/forms/page.mdx

Reviewer: @mj12albert could we apply this change?
Or do you disagree with it?
I could suggest replacing `Integrate Base UI components` with `Integrate components`. It avoids a line break in the ToC and should be sufficiently self-explanatory. 🤷

# Forms

<Subtitle>A guide to building forms with Base UI components.</Subtitle>
<Meta name="description" content="A guide to buildling forms with Base UI components." />

Base UI form control components extend the native [constraint validation API](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#the-constraint-validation-api), to enable building forms that collect user input or provide control over an interface.

TODO: kitchen sink demo here?

## Naming form controls

Form controls must have an accessible name in order to be recognized by assistive technologies. `<Field.Label>` and `<Field.Description>` will automatically assign the accessible name and description to their associated control:

```tsx {3-4, 9-10}
<Form>
  <Field.Root>
    <Field.Label>Time zone</Field.Label>

>     <Field.Description>Used for notifications and reminderss</Field.Description>

Reviewer 2: ```suggestion
    <Field.Description>Used for notifications and reminders</Field.Description>
```

  <Field.Error />
</Field.Root>;
```

## Displaying errors

Use `<Field.Error>` without `children` to automatically display the field's native error message when invalid. The `match` prop can be used to customize the message based on the validity state, and manage internationalization from your application logic:

```tsx title="Customizing error message for a required field"
<Field.Error match="valueMissing">You must create a username</Field.Error>
```

## Submitting data

Once the form passes validation, your app can take over form submission in `onSubmit`:

```tsx {4-9} title="Submitting a form to an external API"
import { Form } from '@base-ui-components/react/form';

<Form
  onSubmit={async (event) => {
    // Prevent the browser's default full-page refresh
    event.preventDefault();
    // Create a FormData object
    const formData = new FormData(event.currentTarget);
    // Process the data, e.g. send the FormData instance in a fetch request
    try {
      await fetch('https://api.example.com', {
        method: 'POST',
        body: formData,
      });
    } catch (error) {
      throw new Error(error);
    }
  }}
/>;
```

## React Hook Form

> [React Hook Form](https://react-hook-form.com) is a popular library that can be integrated with Base UI to externally manage form and field state for your existing components.

Reviewer 3: ```suggestion
[React Hook Form](https://react-hook-form.com) is a popular library that you can integrate with Base UI to externally manage form and field state for your existing components.
```

  }) => (
    <Field.Root name={name} invalid={invalid} touched={isTouched} dirty={isDirty}>
      <Field.Label>Username</Field.Label>
      <Field.Description>
        May appear where you contribute or are mentioned. You can remove it at any time.
      </Field.Description>
      <Field.Control
        placeholder="e.g. alice132"
        value={value}
        onBlur={onBlur}
        onValueChange={onChange}
        ref={ref}
      />
      <Field.Error match={!!error}>
        {error?.message}
      </Field.Error>
    </Field.Root>
  )}
/>
```

### Submitting data

Wrap your submit handler function with `handleSubmit` to receive the form values as a JS object for further handling:

```tsx title="Form submission with React Hook Form"
import { useForm } from 'react-hook-form';
import { Form } from '@base-ui-components/react/form';

interface FormValues {
  username: string;
  email: string;
}

const { handleSubmit } = useForm<FormValues>();

async function submitForm(data: FormValues) {
  // transform the object and/or submit it to a server
  await fetch(/* ... */);
}

<Form onSubmit={handleSubmit(submitForm)} />;
```

## TanStack Form

> [TanStack Form](https://tanstack.com/form/v1/docs/overview) is a newer library also supported by Base UI, and uses a function-based API to orchestrate validations.

Reviewer 3: best practice: avoid mentions of "newness" to make the text more evergreen

```suggestion
Base UI also supports [TanStack Form](https://tanstack.com/form/v1/docs/overview), which uses a function-based API to orchestrate validations.
```

import { useForm } from '@tanstack/react-form';
import { Field } from '@base-ui-components/react/field';

const form = useForm(/* defaultValues, other parameters */)

<form>
  <form.Field
    name="username"
    children={(field) => (
      <Field.Root
        name={field.name}
        invalid={!field.state.meta.isValid}
        dirty={field.state.meta.isDirty}
        touched={field.state.meta.isTouched}
      >
        <Field.Label>Username</Field.Label>
        <Field.Control
          value={field.state.value}
          onValueChange={field.handleChange}
          onBlur={field.handleBlur}
          placeholder="e.g. bob276"
        />

        <Field.Error match={!field.state.meta.isValid}>
          {field.state.meta.errors.join(',')}
        </Field.Error>
      </Field.Root>
    )}
  />
</form>
```

The Base UI `<Form>` component is not needed when using TanStack Form.

### Form validators

To configure a native `<form>`-like validation strategy:

1. Use the additional `revalidateLogic` hook and pass it to `useForm`

> 2. Pass a validation function to the `validators.onDynamic` prop on `<form.Field>` that returns an error object with keys corresponding to the field `name`s

Reviewer 3: ```suggestion
2. Pass a validation function to the `validators.onDynamic` prop on `<form.Field>` that returns an error object with keys corresponding to the field `name`s.
```

# Forms

<Subtitle>A guide to building forms with Base UI components.</Subtitle>

> <Meta name="description" content="A guide to buildling forms with Base UI components." />

Reviewer: Nitpick: Is there a benefit in providing a meta field with the same content as the page subheading?
P.S.: I'd reckon that `Subtitle` should render an `h2` or we should use a slightly different component. 🤔

```tsx {8-9, 14-15} title="Labeling select and slider"
import { Form } from '@base-ui-components/react/form';
import { Field } from '@base-ui-components/react/field';
import { Select } from '@base-ui-components/react/select';
import { Slider } from '@base-ui-components/react/slider';

<Form>
  <Field.Root>
    <Field.Label>Time zone</Field.Label>
    <Field.Description>Used for notifications and reminders</Field.Description>
    <Select.Root />
  </Field.Root>

  <Field.Root>
    <Field.Label>Zoom level</Field.Label>
    <Field.Description>Adjust the size of the user interface</Field.Description>
    <Slider.Root />
  </Field.Root>
</Form>;
```

You can implicitly label `<Checkbox>`, `<Radio>` and `<Switch>` components by enclosing them with `<Field.Label>`:

```tsx title="Implicitly labeling a switch"
import { Field } from '@base-ui-components/react/field';
import { Switch } from '@base-ui-components/react/switch';

<Field.Root>
  <Field.Label>
    <Switch.Root />
    Developer mode
    <Field.Description>Enables extra tools for web developers</Field.Description>
  </Field.Label>
</Field.Root>;
```

> Compose `<Fieldset>` with components that contain multiple `<input>` elements, such as `<CheckboxGroup>`, `<RadioGroup>`, and multi-thumb `<Slider>`s, using `<Fieldset.Legend>` to label the group:

Reviewer: Nitpick: Avoiding "s" after the component name.
```suggestion
Compose `<Fieldset>` with components that contain multiple `<input>` elements, such as `<CheckboxGroup>`, `<RadioGroup>`, and `<Slider>` with multiple Thumbs, using `<Fieldset.Legend>` to label the group:
```

```tsx {5-7} title="Text input using custom asynchronous validation"
import { Field } from '@base-ui-components/react/field';

<Field.Root
  name="username"
  validationMode="onChange"
  validationDebounceTime={300}
  validate={async (value) => {
    if (value === 'admin') {
      /* return an error message when invalid */
      return 'Reserved for system use.';
    }

    const result = await fetch(
      {/* prettier-ignore */},
      /* check the availability of a username from an external API */
    );

    if (!result) {
      return `${value} is unavailable.`;
    }

    /* return `null` when valid */
    return null;
  }}
>
  <Field.Control required minLength={3} />
  <Field.Error />
</Field.Root>;
```

## Displaying errors

Use `<Field.Error>` without `children` to automatically display the field's native error message when invalid. The `match` prop can be used to customize the message based on the validity state, and manage internationalization from your application logic:

```tsx title="Customizing error message for a required field"
<Field.Error match="valueMissing">You must create a username</Field.Error>
```

## Submitting data

> Once the form passes validation, your app can take over form submission in `onSubmit`:

Author: Yeah updated the main demo, and documented both here

  }) => (
    <Field.Root name={name} invalid={invalid} touched={isTouched} dirty={isDirty}>
      <Field.Label>Username</Field.Label>
      <Field.Description>
        May appear where you contribute or are mentioned. You can remove it at any time.
      </Field.Description>
      <Field.Control
        placeholder="e.g. alice132"
        value={value}
        onBlur={onBlur}
        onValueChange={onChange}
        ref={ref}
      />
      <Field.Error match={!!error}>
        {error?.message}
      </Field.Error>
    </Field.Root>
  )}
/>
```

### Submitting data

Wrap your submit handler function with `handleSubmit` to receive the form values as a JS object for further handling:

```tsx title="Form submission with React Hook Form"
import { useForm } from 'react-hook-form';
import { Form } from '@base-ui-components/react/form';

interface FormValues {
  username: string;
  email: string;
}

const { handleSubmit } = useForm<FormValues>();

async function submitForm(data: FormValues) {
  // transform the object and/or submit it to a server
  await fetch(/* ... */);
}

<Form onSubmit={handleSubmit(submitForm)} />;
```

## TanStack Form

> Base UI also supports [TanStack Form](https://tanstack.com/form/v1/docs/overview), which uses a function-based API to orchestrate validations.

Reviewer: WDYT about rephrasing this paragraph to be similar to the `rhf`?
Now it seems that we have some "special sauce" to support this library, when it's only about integration with our components, like with `rhf`. 🤔

> import { DemoTanstackForm } from './demos/tanstack-form';

Author: See https://github.com/mui/base-ui/pull/2948

It's by design on their side, should probably add a small section here with a small code block, for a large form it gets very verbose to do it

> import { DemoTanstackForm } from './demos/tanstack-form';

Reviewer: Oh, gotcha. 👍 
Then, it's only the section headings left, and even those could be left for a follow-up if they need unrelated internal changes. 👌

        errors.email = 'Email is required.';
      } else if (!isValidEmail(formValues.email)) {
        errors.email = 'Invalid email address.';
      }

      return { form: errors, fields: errors };
    },
  },
});
```

### Field validators

You can pass additional validator functions to individual `<form.Field>`s to add validations on top of the form-level validators:

```tsx {8-16} title="Field-level validation with TanStack Form"
import { Field } from '@base-ui-components/react/field';
import { useForm } from '@tanstack/react-form';

const form = useForm();

<form.Field
  name="username"
  validators={{
    onChangeAsync: async ({ value: username }) => {
      const result = await fetch(
        /* check the availability of a username from an external API */
      );

      return result.success ? undefined : `${username} is not available.`
    }
  }}
  children={(field) => (
    <Field.Root name={field.name} /* forward the field props */ />
  )}
>
```

### Submission handling

To submit the form:

1. Pass a submit handler function to the `onSubmit` parameter of `useForm`.
2. Call `form.handleSubmit()` from an event handler such as form `onSubmit` or `onClick` on a button.

> ```tsx {4-7, 13, 17} title="Form submission with TanStack Form" "form.handleSubmit()"

Author: Yeah seems best to omit it everywhere 👍

    fieldState: { invalid, isTouched, isDirty, error },
  }) => (
    <Field.Root name={name} invalid={invalid} touched={isTouched} dirty={isDirty}>
      <Field.Label>Username</Field.Label>
      <Field.Description>
        May appear where you contribute or are mentioned. You can remove it at any time.
      </Field.Description>
      <Field.Control
        placeholder="e.g. alice132"
        value={value}
        onBlur={onBlur}
        onValueChange={onChange}
        ref={ref}
      />
      <Field.Error match={!!error}>
        {error?.message}
      </Field.Error>
    </Field.Root>
  )}
/>
```

### Submitting data

Wrap your submit handler function with `handleSubmit` to receive the form values as a JS object for further handling:

```tsx title="Form submission handler"
import { useForm } from 'react-hook-form';
import { Form } from '@base-ui-components/react/form';

interface FormValues {
  username: string;
  email: string;
}

const { handleSubmit } = useForm<FormValues>();

async function submitForm(data: FormValues) {
  // transform the object and/or submit it to a server
  await fetch(/* ... */);
}

<Form onSubmit={handleSubmit(submitForm)} />;
```

## TanStack Form

> [TanStack Form](https://tanstack.com/form/v1/docs/overview) is a form library with a function-based API for orchestrating validations that can also be integrated with Base UI.

Reviewer: WDYT about linking to the latest doc?
Or do we want to ensure the stability of the example if following majors bring significant BCs?
```suggestion
[TanStack Form](https://tanstack.com/form/latest/docs/overview) is a form library with a function-based API for orchestrating validations that can also be integrated with Base UI.
```

Author: Yeah I think v1 is good for now... I'm sure v2 will have some big changes

# Forms

<Subtitle>A guide to building forms with Base UI components.</Subtitle>
<Meta name="description" content="A guide to buildling forms with Base UI components." />

Base UI form control components extend the native [constraint validation API](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#the-constraint-validation-api) so you can build forms for collecting user input or providing control over an interface. Integrates seamlessly with third-party libraries like [React Hook Form](#react-hook-form) and [TanStack Form](#tanstack-form).

import { DemoBaseUIForm } from './demos/hero';

> <DemoBaseUIForm showExtraPlaygroundLink />

Reviewer 2: Select trigger's and radio group's outline is 1px, not 2px like the other controls
